[kotlin][client] Add inheritance to Kotlin Client templates

mtraynham opened the pull request:
Adds inheritance to the Kotlin Client codegen templates using interfaces as the parent class.

<!-- Please check the completed items below -->
### PR checklist

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [X] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [X] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@jimschubert (2017/09) ❤️, @dr4ke616 (2018/08) @karismann (2019/03) @Zomzog (2019/04) @andrewemery (2019/10)

jimschubert commented on samples/client/petstore/kotlin/src/main/kotlin/org/openapitools/client/models/Order.kt:
```diff
@@ -38,8 +38,12 @@ data class Order (
+
+
+
+ : Serializable
```
I think we'll need to clean up the newly added empty lines before merge.

mtraynham replied:
Corrected by putting all import variants on the same line.

jimschubert commented on modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:
```diff
@@ -46,6 +52,14 @@ import java.io.Serializable
+{{#optionalVars}}
+{{>interface_opt_var}}
+{{/optionalVars}}
+{{/discriminator}}
```
Not tested, but I think this would need to be the following to avoid empty newlines:

```
{{/serializableModel}}{{#discriminator}}{{#requiredVars}}
{{>interface_req_var}}{{/requiredVars}}{{#optionalVars}}
{{>interface_opt_var}}{{/optionalVars}}{{/discriminator}}
```

mtraynham replied:
Corrected, thanks!

jimschubert commented:
@mtraynham thanks for keeping updated with master. That's much appreciated.  My only concern about the newly added code is the empty newlines it introduces for noop conditions.

I wonder if we can target 4.3.x with this new code? I have concerns that the introduction of the new template files may cause issue, and just to be safe we could release in the minor version which is "breaking change with fallback". The fallback being users could use 4.2.x templates if they have issues.

Any concern with retargeting? As an alternative we could get the vendor extension into master and the template changes against the 4.3.x branch.

mtraynham commented:
Thanks @jimschubert for taking a look!  I'm not against flattening the new lines, it just makes the template kinda of unreadable.  I'll rebase against the 4.3.x branch as well.

mtraynham commented:
@jimschubert , I corrected the issues you mentioned and rebased my commits on to OpenAPITools/4.3.x.  

Looks like the checks are failing because it's attempting to merge the feature branch into master and getting some merge failures.  I can rebase again when those are corrected if need be.

jimschubert commented:
Verified inheritance after testing a merge of master. Looks good!
